src/for_mathlib/subring.lean

```diff
@@ -0,0 +1,43 @@
+import for_mathlib.add_subgroup group_theory.submonoid
+import poly
+import algebra.ring
+local attribute [instance] classical.prop_decidable
+
+open group
+
+variables {R : Type} [ring R]
+
+/-- `S` is a subring: a set containing 1 and closed under multiplication, addition and and additive inverse. -/
+class is_subring  (S : set R) extends is_add_subgroup S, is_submonoid S : Prop.
+
+instance subset.ring {S : set R} [is_subring S] : ring S :=
+{ add_comm      := assume ⟨a,_⟩ ⟨b,_⟩, subtype.eq $ add_comm _ _,
+  left_distrib  := assume ⟨a,_⟩ ⟨b,_⟩ ⟨c,_⟩, subtype.eq $ left_distrib _ _ _,
+  right_distrib := assume ⟨a,_⟩ ⟨b,_⟩ ⟨c,_⟩, subtype.eq $ right_distrib _ _ _,
+  .. subtype.add_group,
+  .. subtype.monoid }
+
+instance subtype.ring {S : set R} [is_subring S] : ring (subtype S) :=
```

Comment:
My understanding from Chris is that this instance should be dumped and not used, and any time it was triggering in our repo we should remove the explicit invocation of `subtype` and let the coercion do the work. If you remove it, can you fix everything that breaks?

Reply:
So, the first thing that breaks is line 29, just below it: proving that `subtype.val` is a ring hom. I can replace it with

Reply:
My understanding from Mario now is that we should be using both instances, but one should be proved using the other. He said "That's not quite what I'm saying. You should have both instances, but you can prove one by just referencing the other".